feat: Make toolbox & flyout properly focusable.

BenHenning opened the pull request:
Fixes #180
Fixes #149
Fixes part of #101

The toolbox already had a tab stop, but it wouldn't become properly enabled for keyboard navigation. Now it is properly enabled and works when focused both via tab navigation and with the 'T' shortcut.

This has fairly substantial behavioral differences now:
- Focus has been reworked to logically support _either_ the toolbox _or_ the workspace currently having DOM focus in order to support the toolbox properly having keyboard navigation support when enabled.
- Keyboard navigation provided by core Blockly for toolbox has been monkey patched out since it would otherwise interfere with the toolbox navigation (and results in items being skipped when navigating through the categories).
- The navigation controller now has support for a 'nowhere' state to support blurring the toolbox.
- It's possible for the toolbox to have keyboard input with or without DOM focus which can result in some quirkiness when tabbing between the two--this may be something to iterate on if it introduces any user confusion.

From a long-term perspective, we may want to:
- Limit some shortcuts to only cases when the workspace is focused (there's a behavior difference with either approach compared to the currently implementation).
- Figure out a good way to ensure that these behaviors are properly captured in core Blockly as this plugin is moved over.

Separately, note that:
- Per feedback from @rachel-fenichel this only handles toolbox focus and ignores flyouts altogether. Proper flyout support can be added later as needed.
- A subcategory was added to the demo toolbox categories as part of the analysis from #149 (see the second video below).

The new behavior can be seen here:

[Screen recording 2025-03-11 4.59.00 PM.webm](https://github.com/user-attachments/assets/b123d90d-69fb-42a7-b3f6-3ba173392264)

Note that the specific behaviors to observe are around how tabbing to the toolbox allows it to be focused and have keyboard navigation enabled. The video also demonstrates some of the awkward behaviors of switching focus between the toolbox and workspace when the other one is actually focused (such as when inserting a block from a focused toolbox or using the 't' command from an active workspace).

Note that this video also shows how subcategories are currently a bit buggy with left/right navigation (an existing issue now being tracked in #241):

https://github.com/user-attachments/assets/90b89b0d-0242-4863-8414-b4383b68c8e5

BenHenning committed:
feat: Make toolbox & flyout properly focusable.

The toolbox already had a tab stop, but it wouldn't become properly
enabled for keyboard navigation. Now it is properly enabled and works
when focused both via tab navigation and with the 'T' shortcut.

rachel-fenichel commented on src/index.ts:
```diff
@@ -82,14 +82,26 @@ export class KeyboardNavigation {
     workspace.getSvgGroup().addEventListener('focus', this.focusListener);
     workspace.getSvgGroup().addEventListener('blur', this.blurListener);
+
+    const toolbox = workspace.getToolbox();
```
Also remove these in `dispose` below.

BenHenning replied:
Done.

rachel-fenichel commented on src/navigation_controller.ts:
```diff
@@ -148,33 +159,65 @@ export class NavigationController {
+  updateToolboxOrFlyoutFocus(workspace: WorkspaceSvg, isFocused: boolean) {
     if (isFocused) {
-      this.navigation.focusWorkspace(workspace, true);
+      if (!workspace.getToolbox()) {
```
nit: I would flip this to
```
if (workspace.getToolbox()) {
  this.navigation.focusToolbox(workspace);
} else {
  this.navigation.focusFlyout(workspace);
}
```

Fortunately none of the workspace we're testing with have neither a flyout nor a toolbox.

rachel-fenichel replied:
And see comment in `src/index.ts` about potentially just making this handle the toolbox case and skipping the flyout case if you haven't tested it.

BenHenning replied:
Flipped the logic at first, but then changed it a bit more due to the other comment and removing explicit flyout support.

rachel-fenichel commented on src/navigation_controller.ts:
```diff
@@ -148,33 +159,65 @@ export class NavigationController {
+    if (isFocused) {
+      this.navigation.focusWorkspace(workspace, true);
+      this.navigationFocus = NAVIGATION_FOCUS_MODE.WORKSPACE;
+    } else this.navigationFocus = NAVIGATION_FOCUS_MODE.NONE;
```
Style: Please use brackets even for a one-liner. We only skip them for `if (x) return;`

BenHenning replied:
Ah, I was unsure on this case. :) Fixed.

rachel-fenichel commented on src/navigation.ts:
```diff
@@ -537,14 +537,34 @@ export class Navigation {
+      case Constants.STATE.FLYOUT:
+      case Constants.STATE.TOOLBOX:
+        // Clear state since neither the flyout nor toolbox are focused anymore.
+        this.setState(workspace, Constants.STATE.NOWHERE);
```
Will this ever be called when state needs to move directly to the workspace, instead of to `NOWHERE`?

BenHenning replied:
Not directly from what I've observed. The logic here is largely relying on the focus/blur events to be handled in the correct order at the DOM level. When tab navigating from a focused toolbox to the workspace this method will be called to reset state to nowhere, then the workspace will receive focus and this state will immediately be updated to be workspace.

We needed an intermediary state because the assumption that workspace is always 'next' is also false (when shift tabbing from the toolbox).

I'm open to suggestions on alternate designs or ways of making this flow a bit clearer since I recognize that it is a bit obscured currently.

rachel-fenichel replied:
> The logic here is largely relying on the focus/blur events to be handled in the correct order at the DOM level. When tab navigating from a focused toolbox to the workspace this method will be called to reset state to nowhere, then the workspace will receive focus and this state will immediately be updated to be workspace.

Got it. Good to keep in mind as a constraint, in case we later find that the events are coming in in a different order in some configurations, but this explanation makes sense to me.

rachel-fenichel commented on src/navigation_controller.ts:
```diff
@@ -44,6 +45,16 @@ interface Scope {
+  /** Indicates either the toolbox or a flyout has focus. */
+  TOOLBOX_OR_FLYOUT = 'toolbox_or_flyout',
+  /** Indicates that the main workspace currently has focus. */
+  WORKSPACE = 'workspace',
```
Handy since there's a chance we'll need to add `modal` or `widgetdiv` or similar as options.

rachel-fenichel commented on src/index.ts:
```diff
@@ -82,14 +82,26 @@ export class KeyboardNavigation {
+    const toolbox = workspace.getToolbox();
+    if (toolbox != null && toolbox instanceof Blockly.Toolbox) {
+      const contentsDiv = toolbox.HtmlDiv?.querySelector('.blocklyToolboxContents');
+      contentsDiv?.addEventListener('focus', () => {
```
You've tried to generalize this to handle either toolbox or flyout, but does this selector actually work in the case where there is a flyout with no toolbox?

We're only testing with category toolboxes so I'm also fine with this code just setting focus on the toolbox and not trying to handle flyouts--would rather have that instead of generalized code that doesn't work in one of two cases.

BenHenning replied:
I was mostly aiming to mimic existing logic around how the toolbox was handled elsewhere (such as in `openToolboxOrFlyout`), but I have not tested the case of a flyout with no toolbox. I think it makes sense to simplify the solution here to keep it specific to toolboxes--done.

rachel-fenichel reviewed (commented):
This makes sense to me logically.

In the screencast I see that the text on all blocks in the flyout is highlighted as though selected. Please fix.

BenHenning committed:
chore: address reviewer comments

BenHenning committed:
Merge branch 'main' into make-toolbox-properly-focusable

Conflicts:
	src/navigation_controller.ts

BenHenning committed:
feat: add subcategory to demo toolbox categories

This helps to verify subcategory behavior (which has been observed as
not completely working currently).

BenHenning reviewed (commented):
Self-reviewed latest changes.

Thanks @rachel-fenichel. PTAL.

rachel-fenichel commented on src/toolbox_monkey_patch.js:
```diff
@@ -0,0 +1,12 @@
+
+import * as Blockly from 'blockly/core';
+
+Blockly.Toolbox.prototype.onKeyDown_ = function () {
```
In this PR or in a followup (file an issue for it), you should make this only happen on install/uninstall instead of happening on load.

BenHenning replied:
Done, though just to check installation is a bit awkward since it needs to happen even before NavigationController is created--does this seem reasonable?

rachel-fenichel reviewed (approved):
LGTM. Thanks for simplifying out the flyout case for now--please also file a bug to follow up on the flyout case, if we don't already have one.

BenHenning committed:
Address reviewer comment for monkey patch install.

BenHenning committed:
Merge branch 'main' into make-toolbox-properly-focusable

Conflicts:
	src/navigation_controller.ts

rachel-fenichel commented on test/index.ts:
```diff
@@ -83,6 +84,7 @@ function createWorkspace(): Blockly.WorkspaceSvg {
     renderer,
   };
   const blocklyDiv = document.getElementById('blocklyDiv')!;
+  ToolboxMonkeyPatch.install();
```
Oh, I didn't mean to move this all the way up into `test/index.ts`. Instead, call `install` from `NavigationController.init()` and `uninstall` from `NavigationController.dispose()`

rachel-fenichel replied:
However, I think that I'm distracting from the point of this PR by focusing on where that code goes, and it's fine to merge as is.

BenHenning replied:
Reverted the logic back to the way it was before since installation is messy here, and the monkey patches can go away soon (once v12 is available).

BenHenning committed:
Merge branch 'main' into make-toolbox-properly-focusable

Conflicts:
	src/navigation_controller.ts

BenHenning committed:
Revert monkeypatch installation changes.

BenHenning committed:
Add fix for null current node.

It seems possible now to call hide() earlier than previously expected
(i.e. when there's a cursor but no current node, such as in the case of
focusing the toolbox before the workspace as this PR now makes
possible).

BenHenning reviewed (commented):
Self-reviewed latest changes.

@rachel-fenichel thanks for the extra details on the text issue. It seems that it mysteriously went away after syncing to latest, so that's fun. :) If you could, PTAL at the latest changes just to make sure everything seems reasonable. I've enabled auto-merge so feel free to merge this, as well, if there are no follow-ups. The first video in the description has also been updated accordingly.
